### PR TITLE
Unexpected Warning Capture and Message for try_redcapr

### LINKED ISSUE #121: [BUG] Unexpected Warning Related to readr/vroom
# Expected Behavior

REDCapTidieR should provide useful/helpful error/warning messages.

# Current Behavior

A warning came up when using REDCapTidieR on the CHOP BMT Outcomes REDCap Database:

```r
Warning: One or more parsing issues, call `problems()` on your data frame for details,
e.g.:
   dat <- vroom(...)
   problems(dat)
```

It's possible that our outstanding PRs to capture unexpected warnings will mask/resolve this, but we should look into the reason why this is occurring for two reasons:

- Something in the database is weird (a poorly formatted date, or other)
- An unexpected error is not being caught and should be

**Screenshots**
image

### Checklist
Before submitting this issue, please check and verify below that the submission meets the below criteria:

- [x] The issue is atomic
- [x] The issue description is documented
- [x] The issue title describes the problem succinctly
- [x] Developers are assigned to the issue
- [x] Labels are assigned to the issue

### PR DESCRIPTION
**Pull request by rsh52**
# Description

This PR adds an additional unexpected warning capture to our current `try_redcapr` `try_fetch` block. In order to correctly return the REDCapR export and ensure no duplicate warnings are returned, a combination of `cnd_muffle` and `zap` are used.

# Proposed Changes 
List changes below in bullet format:

- Add a `warning` function to the `try_fetch` block
- Write an associated test

**Example with simple parsing error:**

```r
>   try_redcapr(list(success = TRUE, data = as.numeric("A"))) 
[1] NA
Warning message:
! The REDCapR export operation produced an unexpected warning.
Caused by warning:
! NAs introduced by coercion 
```

**Example with BMT Outcomes Database where known issue exists:**

```r
> bmt_out <- read_redcap(redcap_uri, token = token)
Warning message:                                                                                                                
In read_redcap(redcap_uri, token = token) :
  ! The REDCapR export operation produced an unexpected warning.
Caused by warning:
! One or more parsing issues, call `problems()` on your data frame for details, e.g.:
  dat <- vroom(...)
  problems(dat)
```

### Issue Addressed
**Closes #121**

### PR Checklist
Before submitting this PR, please check and verify below that the submission meets the below criteria:

- [x] New/revised functions have associated tests
- [NA] New/revised functions that update downstream outputs have associated static testing files (`.RDS`) updated under `inst/testdata/create_test_data.R`
- [NA] New tests that make API calls use `httptest::with_mock_api` and any new mocks were added to `tests/testthat/fixtures/create_httptest_mocks.R`
- [x] New/revised functions use appropriate naming conventions
- [x] New/revised functions don't repeat code
- [x] Code changes are less than **250** lines total
- [x] Issues linked to the PR using [GitHub's list of keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] The appropriate reviewer is assigned to the PR
- [x] The appropriate developers are assigned to the PR
- [x] Pre-release package version incremented using `usethis::use_dev_version()`

# Code Review
This section to be used by the reviewer and developers during Code Review after PR submission

### Code Review Checklist

- [ ] I checked that new files follow naming conventions and are in the right place
- [ ] I checked that documentation is complete, clear, and without typos
- [ ] I added/edited comments to explain "why" not "how"
- [ ] I checked that all new variable and function names follow naming conventions
- [ ] I checked that new tests have been written for key business logic and/or bugs that this PR fixes
- [ ] I checked that new tests address important edge cases

### CODE REVIEW

**Comment by skadauke** [comment]
Should we call these warnings "unexpected"? I think we call errors "unexpected" since we try to anticipate most/all failure modes. We're not doing that for warnings. Would change message "The REDCapR export operation produced an unexpected warning." to "One of the REDCapR operations produced a warning. See below for details."

**Comment by rsh52** [comment]
> Should we call these warnings "unexpected"? I think we call errors "unexpected" since we try to anticipate most/all failure modes. We're not doing that for warnings. Would change message "The REDCapR export operation produced an unexpected warning." to "One of the REDCapR operations produced a warning. See below for details."

Sounds good to me, implemented in recent commit. 👍

File: R/utils.R [hunk 1]
```diff
@@ -528,6 +528,18 @@ try_redcapr <- function(expr, call = caller_env()) {
...
+
+  # URI and token we want are in the env associated with expr
+  warning$redcap_uri <- env$redcap_uri
+  warning$token <- env$token
+
+  # Defaults for other error components
+  warning$message <- c("!" = "The {.pkg REDCapR} export operation produced an unexpected warning.")
+  warning$class <- "REDCapTidieR_cond"
+  warning$call <- call
+
```
Comment (ezraporter) [inline R/utils.R:542]: I actually don't think we need to duplicate this from L518-L529 since most of the defaults are the same. I would change all the `error$` to `condition$` just so the naming is consistent and then have the `cli_warn()` call on L567 be:

```r
cli_warn(
  c("!" = "What Stephan suggested ;)"),
  call = condition$call,
  parent = cnd,
  class = c("unexpected_warning", condition$class)
)
```

**Review by ezraporter — CHANGES_REQUESTED** [review]
One idea to reduce the number of objects floating around but looks good overall!

File: R/utils.R (hunk 1 above)
Comment (rsh52) [reply R/utils.R:542]: That's fine with me, I chose this on the off chance that down the line we want to throw more specific warnings and increase the list options. But I like simplicity better.

**Review by ezraporter — APPROVED** [review]
